### ISSUE TITLE
[REQ] Image "restoration" tools

### ISSUE THREAD

**Issue by forart** [shirooo39/MiXLab#15]
It could be very interesting to have those cool tools in ML, even to make comparisons !

- https://github.com/tducvu/PIC16B-project
- https://github.com/animalecs/deep_koalarization
- https://github.com/jantic/DeOldify
- https://github.com/shatadal88/Simple-Old-photo-restore-
- https://github.com/flynneastwood/VintagePhotoRestaurator
- https://github.com/awarischool/photo-colorization
- https://github.com/Sarthakh/Automatic-Image-Colorization-Using-Auto-encoders-without-Image-Compression
- https://github.com/AmitNikhade/colorvibes
- https://github.com/Saaketh-J/NewColor
- https://github.com/KorneliuszDyszczynski/ColorizingPhotosNeuralNetwork
- https://github.com/Oleg007003/ColorizationPicturesNetwork
- https://github.com/krisbitney/ImgBuff
- https://github.com/srijag2700/Colorization
- https://github.com/tejasbana/Colorizing_images
- ...

Hope that inspires !

**Comment by shirooo39** [comment]
I've just added (an extemely simple, rough, and unfinished) implementation of <a href="https://colab.research.google.com/github/shirooo39/MiXLab/blob/master/MiXLab.ipynb#scrollTo=Jbw2QIUB6JKR&uniqifier=1" target="_blank">Real-ESRGAN</a>

I'll... think about adding some more later

**Comment by forart** [comment]
Just tested, and looks quite good:
![gundam](https://user-images.githubusercontent.com/171307/135242392-2de5845d-fe41-41d4-a18a-406be151f028.png)

Do you think is possible to adapt it for videos too ? 
(something like: input video -> FFMPEG decoding -> still images -> Real-ESRGAN -> FFMPEG concat/encoding -> restored/upscaled video)

Thanks !

**Comment by shirooo39** [comment]
I think that could be possible, but I know nothing about ffmpeg.
there's already ffmpeg in my notebook (that I took frome somewhere else) and maybe you can use it to extract each frames, use Real-ESRGAN to upscale them all, and then use ffmpeg again to put them all back together.

though I can't imagine how long that'd take

**Comment by forart** [comment]
Well, the main "problem" could be more the disk space than rendering time...

Anyway, just found this interesting notebook by @tg-bomze that may inspire:
https://colab.research.google.com/github/tg-bomze/ENTAR/blob/master/ENTAR_Eng.ipynb

**Comment by forart** [comment]
Clean SwinIR wrapper by @Lin-Sinorodin: https://github.com/Lin-Sinorodin/SwinIR_wrapper

**Comment by forart** [comment]
Very interesting find (thanks to Selur, the [Hybrid](https://www.selur.de/) author): [HINet - Half Instance Normalization Network for Image Restoration](https://github.com/megvii-model/HINet) by @megvii-model
>With the help of HIN Block, HINet surpasses the state-of-the-art (SOTA) on various image restoration tasks. For image denoising, we exceed it 0.11dB and 0.28 dB in PSNR on SIDD dataset, with only 7.5% and 30% of its multiplier-accumulator operations (MACs), 6.8 times and 2.9 times speedup respectively. For image deblurring, we get comparable performance with 22.5% of its MACs and 3.3 times speedup on REDS and GoPro datasets. For image deraining, we exceed it by 0.3 dB in PSNR on the average result of multiple datasets with 1.4 times speedup.

**Comment by forart** [comment]
According to @Lin-Sinorodin:

> Thanks for the comment, I'm glad you found it useful!
> 
> Although it looks very interesting, I won't have the time to work on this in the near future...
> 
> If you are willing to try it out yourself, it seems from their code that it is pretty straightforward. If you look at the code on [this page](https://github.com/megvii-model/HINet/blob/main/basicsr/demo.py), It seems that something like this can be a pretty clean and short wrapper:
> 
> ```python
> import cv2
> import torch
> 
> from basicsr.models import create_model
> from basicsr.train import parse_options
> from basicsr.utils import FileClient, imfrombytes, img2tensor, padding
> 
> # define model
> opt = parse_options(is_train=False)
> model = create_model(opt)
> 
> img_lq = cv2.imread(path, cv2.IMREAD_COLOR)
> img_lq = cv2.cvtColor(img_lq, cv2.COLOR_BGR2RGB)
> 
> img_lq_tensor = torch.from_numpy(img_lq.transpose(2, 0, 1)).float()
> 
> # get high res img
> img_hq: np.array = model.my_single_image_inference(img_lq_tensor , output_path)
> ```
> 
> The main difference is adding a function similar to the [inference function](https://github.com/megvii-model/HINet/blob/033ed3a5470950b926b7437c0994a8665fba5ab5/basicsr/models/image_restoration_model.py#L269) that return the image instead of saving it:
> 
> ```python
> def my_single_image_inference (self, img) -> np.array:
>     self.feed_data(data={'lq': img.unsqueeze(dim=0)})
> 
>     if self.opt['val'].get('grids', False):
>         self.grids()
> 
>     self.test()
> 
>     if self.opt['val'].get('grids', False):
>         self.grids_inverse()
> 
>     visuals = self.get_current_visuals()
>     sr_img = tensor2img([visuals['result']])
> 
>     # REPLACE THIS
>     # imwrite(sr_img, save_path)
> 
>    # WITH THIS
>    return sr_img
> ```
> 
> It's just in a glance, and I probably missed something but the main idea is here. This way it's should be also available to merge this into a more high-level API, that supports both SwinIR and the paper you referenced. Seems like a cool project, and I'll be happy to contribute.
> 
> Hope it's helpful, Lin
